### PR TITLE
[ZEPPELIN-1258] Add Spark packages support to Livy interpreter

### PR DESCRIPTION
**Pull request by mfelgamal**
### What is this PR for?
Adding extra libraries to livy interpreter which isn't exist by default.

### What type of PR is it?
[ Improvement ]

### Todos
* [Test case ] - Task

### What is the Jira issue?
*  [ZEPPELIN-1258]

### How should this be tested?
- Create new livy interpreter or modify the default.
- Set `livy.spark.jars.packages` to list of maven coordinates of jars. The format for the coordinates should be groupId:artifactId:version. 

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? yes

### CODE REVIEW

**Comment by zjffdu** [comment]
I would not recommend user to do that to add extra library. Because most of time zeppelin user don't know the cluster details like which extra library are located in the cluster machines. I would suggest livy to support dynamic library loading as native spark interpreter.

**Comment by jimdowling** [comment]
@zjffdu - I don't think there's a ZeppelinContext available in Livy. Would this be acceptable until the ZeppelinContext becomes available? As of now, Livy is not that much use if you cannot add libraries to your Spark app.

**Comment by zjffdu** [comment]
@jimdowling  Right, there's no ZeppelinContext in livy right now. Would allowing user to specify jars/packages when creating livy session more general ? (like --jars/--packages in spark-submit)

**Comment by jimdowling** [comment]
Do you mean by modifying the UI on notebook creation or using a param to selecting the interpreter like
 '%livy --jars [path]

**Comment by zjffdu** [comment]
No, I mean add interpreter property to allow user to specify jars/packages.

**Comment by mfelgamal** [comment]
I think that the livy itself doesn't support add  (--jars/--packages property in spark-submit)

**Comment by zjffdu** [comment]
You can use --conf spark.jars=<jars> and --conf spark.jars.packages=<packages> instead

**Comment by mfelgamal** [comment]
@zjffdu I added `spark.jars.packages` property in the recent commit, and it works correctly.

**Comment by mfelgamal** [comment]
@zjffdu do you have any further comments ?

File: docs/interpreter/livy.md [hunk 1]
```diff
@@ -102,8 +102,31 @@ Example: `spark.master` to `livy.spark.master`
...
 
+## Adding External libraries
+You can load dynamic library to livy interpreter by set `livy.spark.jars.packages` property to comma-separated list of maven coordinates of jars to include on the driver and executor classpaths. The format for the coordinates should be groupId:artifactId:version. 
+
+Example
+
+<table class="table-configuration">
+  <tr>
+    <th>Property</th>
+    <th>Default</th>
```
Comment (felixcheung) [inline docs/interpreter/livy.md]: this wouldn't be the "Default" right?

**Comment by felixcheung** [comment]
could you update the title of this PR and JIRA ZEPPELIN-1258, if we are changing how we would approach this?

File: docs/interpreter/livy.md (hunk 1 above)
Comment (mfelgamal) [inline docs/interpreter/livy.md]: True, it's just example.

**Comment by mfelgamal** [comment]
@felixcheung I suggests this title "Adding packages to livy interpreter", Is it Ok ?

**Comment by felixcheung** [comment]
how about "Add Spark packages support to Livy interpreter"?

**Comment by mfelgamal** [comment]
@felixcheung done.

File: docs/interpreter/livy.md (hunk 1 above)
Comment (felixcheung) [inline docs/interpreter/livy.md]: Let's change it to Example?

Comment (mfelgamal) [inline docs/interpreter/livy.md]: @felixcheung done.

**Comment by mfelgamal** [comment]
@felixcheung  do you have any further comments ?

**Comment by felixcheung** [comment]
There is a test failure?

**Comment by mfelgamal** [comment]
@felixcheung  : This PR does not change  the code. It is just a change in the configurations. So, it should not affect the tests

**Comment by felixcheung** [comment]
That's a fair point.

I'll merge if there is no more discussion.

**Pull request closed without merging**